### ISSUE TITLE
USB Hub, Wireless, Printer, etc.

### ISSUE THREAD

**Issue by Torinde** [crazii/USBDDOS#1]
Is it possible to add support for ...?
- USB hub
- Wireless Controller - primarily for Bluetooth (mouse, keyboard, [audio](https://github.com/crazii/SBEMU/issues/39), etc.), maybe also RNDIS
- Printer
- more CDC devices - primarily UART/RS-232, then Ethernet and WiFi, maybe also modems
- parallel port (also a CDC device?)
- Scanner 
- Physical interface device (PID) Force feedback joystick - is there DOS software/game that supports force feedback?

**Comment by crazii** [comment]
Theoretically it is possible, but..., by the current state of USBDDOS (which is not stable and only works in VirtualBox), it is very hard for now.
I've tested it on some real machine, there are various problems (i.e. controller not found, USB mouse emulation behaving wrong with DOS games).

**Comment by crazii** [comment]
Theoretically it is possible, but..., by the current state of USBDDOS (which is not stable and only works in VirtualBox), it is very hard for now.
I've tested it on some real machine, there are various problems (i.e. controller not found, USB mouse emulation behaving wrong with DOS games).

**Comment by crazii** [comment]
My network is f**ked and cannot post comment, sorry if I cannot reply you in time.

**Comment by Torinde** [comment]
OK, I understand. I've seen [mentioned an example set of USB drivers for DOS - USBUHCI, USBMOUSE, CuteMouse](https://retrocomputing.stackexchange.com/questions/17700/use-a-usb-mouse-for-dos-games) - per the description at [Bret Johnson's page](https://bretjohnson.us) it includes source code and UHCI drivers for keyboards, mice, joysticks, printers, disk drives, and hubs, as well as several different test and support programs. Will that help?

Another use-case are external USB-to-ISA adapters [1](http://arstech.com/install/ecom-prodshow/usb2isar.html) [2](https://www.arstech.com/install/cms-display/ste_usl.html) - they even sell [dosbox-x.conf and win9x.img](https://arstech.com/install/index.php?app=ecom&ns=prodshow&ref=img-wME-w98) files! [I wonder](https://github.com/joncampbell123/dosbox-x/issues/3269#issuecomment-1037106568) how they get access to the USB-attached ISA slots in DOSbox-X - via .conf chages, by some DOS driver, by changing the DOSbox-X source code itself?

**Comment by crazii** [comment]
Yes that helps, actually some of the USBDDOS code has the same implementations (maybe in different form, inlined asm  or C etc) because I learned from it :D. there's also comment in the source code that have mentioned it or have links.

**Comment by Torinde** [comment]
FYI, [USB in DOS](http://pmwiki.xaver.me/drdoswiki/index.php?n=Main.USB) - I don't see anything directly applicable to this project, but maybe there is something (it has general information, mentions the USBDOS from above, a commercial/closed source DOSUSB product, links to some USB device drivers for DOS)

**Comment by jan-kleks** [comment]
@crazii It would have been lovely to have support for Xbox controllers (it is probably the most popular gamepad for PCs). With gamepads/joysticks support is in (potentially simulating a keyboard/mouse if necessary) + SBEMU, and modern DOS gaming would become so much easier...

**Comment by crazii** [comment]
Forgive me for my ignorance, I never heard of FAT28, is it a FAT file format? or related 
Also I didn't implement INT 13 hook in the BIOS level, just a install & modify of DOS drivers to handle disk IO, I don't know whether it is enough. Let me know if there's any test cases.
Callback address rejection - I didn't know that it stands for either, please help me clarify. Any links will help.

I thought about hotplugging, it can be done, but the hardest part is hot-unplugging, there's too much cases to cause errors, and unrecoverable errors.
USBDDOS loads itself in the command line, not in CONFIG.SYS (not supported), but there's no unloading. The worst cases of unloading is like unplugging, i.e. usbddos gained control of drive C (a booting USB disk) and unloading usbddos will make the system unable to run.

**Comment by crazii** [comment]
> Low speed hubs connected to high speed hubs - that's mentioned as one of the TODO items for DOSUSB, along with Isochronous transfer, thus it warrants at least testing/confirmation.

I just read the document of DOSUSB, the HID on high speed hub is still in TODO. HIDs usually are low speed devices (the one I tested with is low speed, bought in recent 2 years), so both the HID & low speed hub is the same thing: low speed devices on high speed hub (at least for EHCI to use high speed hub).
I've done that and tested the HID mouse, it's working, so it's very likely that low speed hubs are working too.

**Comment by crazii** [comment]
NO, that probably needs remap GPT partition info into MBR before hand it to DOS, not done.
